File: packages/modules/chess/ts/src/notation/notation.ts
import type {
  Board,
  Color,
  CastlingRights,
  Square,
  Move,
  PieceType,
  GameState,
} from '../types/types';
import { fromFenBoard, toFenBoard } from '../board/board';
import { getLegalMoves, getPseudoLegalMoves } from '../moves/moves';
import { isInCheck } from '../moves/attack';
import {
  getRank,
  getFile,
  toSquare,
  toSquareName,
  toSquareFromName,
} from '../utils/utils';

export const fromFen = (fen: string): GameState => {
  const parts = fen.split(' ') as [
    string,
    string,
    string,
    string,
    string,
    string,
  ];

  const board = fromFenBoard(parts[0]!);

  const turn: Color = parts[1] === 'b' ? 'b' : 'w';

  const castlingRights: CastlingRights = {
    K: false,
    Q: false,
    k: false,
    q: false,
  };
  if (parts[2] !== '-') {
    for (const ch of parts[2]!) {
      if (ch === 'K') castlingRights.K = true;
      else if (ch === 'Q') castlingRights.Q = true;
      else if (ch === 'k') castlingRights.k = true;
      else if (ch === 'q') castlingRights.q = true;
    }
  }

  const enPassant = parts[3] === '-' ? null : toSquareFromName(parts[3]!);

  const halfMoveClock = parseInt(parts[4]!) || 0;
  const fullMoveNumber = parseInt(parts[5]!) || 1;

  return {
    board,
    turn,
    castlingRights,
    enPassant,
    halfMoveClock,
    fullMoveNumber,
    history: [],
    status: 'playing',
    result: '*',
    inCheck: isInCheck(board, turn),
  };
};

export const toFen = (state: GameState): string => {
  const fen = toFenBoard(state.board);

  const cr = state.castlingRights;
  let castling = '';
  if (cr.K) castling += 'K';
  if (cr.Q) castling += 'Q';
  if (cr.k) castling += 'k';
  if (cr.q) castling += 'q';

  return `${fen} ${state.turn} ${castling || '-'} ${state.enPassant !== null ? toSquareName(state.enPassant) : '-'} ${state.halfMoveClock} ${state.fullMoveNumber}`;
};

const PIECE_LETTER: Record<PieceType, string> = {
  p: '',
  n: 'N',
  b: 'B',
  r: 'R',
  q: 'Q',
  k: 'K',
};

export const toUci = (move: Move): string =>
  toSquareName(move.from) + toSquareName(move.to) + (move.promotion ?? '');

export const fromUci = (uci: string): Move | null => {
  if (uci.length < 4) return null;
  const from = toSquareFromName(uci.slice(0, 2));
  const to = toSquareFromName(uci.slice(2, 4));
  if (from === null || to === null) return null;
  const promoChar = uci[4] as PieceType | undefined;
  const promotion =
    promoChar && ['q', 'r', 'b', 'n'].includes(promoChar) ? promoChar : null;
  return { from, to, promotion, captured: null };
};

export const toSan = (
  board: Board,
  move: Move,
  turn: Color,
  castlingRights: CastlingRights,
  enPassant: Square | null
): string => {
  const piece = board[move.from]!;
  const fileDiff = Math.abs(getFile(move.to) - getFile(move.from));

  if (piece.type === 'k' && fileDiff === 2) {
    return getFile(move.to) > getFile(move.from) ? 'O-O' : 'O-O-O';
  }

  if (piece.type === 'p') {
    const capture = fileDiff !== 0 || move.captured;
    const promo = move.promotion ? `=${move.promotion!.toUpperCase()}` : '';
    const file = capture ? toSquareName(move.from)[0]! : '';
    const capInd = capture ? 'x' : '';
    return `${file}${capInd}${toSquareName(move.to)}${promo}`;
  }

  const letter = PIECE_LETTER[piece.type];

  const others = getPseudoLegalMoves(
    board,
    turn,
    castlingRights,
    enPassant
  ).filter((m) => {
    const p = board[m.from];
    return (
      p &&
      p.type === piece.type &&
      m.to === move.to &&
      m.from !== move.from &&
      getLegalMoves(board, turn, castlingRights, enPassant).some(
        (lm) => lm.from === m.from && lm.to === m.to
      )
    );
  });

  let disambig = '';
  if (others.length > 0) {
    const sameFile = others.some((m) => getFile(m.from) === getFile(move.from));
    const sameRank = others.some((m) => getRank(m.from) === getRank(move.from));
    if (sameFile) disambig = toSquareName(move.from);
    else if (sameRank) disambig = toSquareName(move.from)[0]!;
    else disambig = toSquareName(move.from)[0]!;
  }

  const capture = move.captured ? 'x' : '';
  return `${letter}${disambig}${capture}${toSquareName(move.to)}`;
};

export const fromSan = (
  san: string,
  board: Board,
  turn: Color,
  castlingRights: CastlingRights,
  enPassant: Square | null
): Move | null => {
  const moves = getLegalMoves(board, turn, castlingRights, enPassant);

  if (san === 'O-O' || san === 'O-O-O') {
    const rank = turn === 'w' ? 0 : 7;
    const kingFile = san === 'O-O' ? 6 : 2;
    return (
      moves.find((m) => {
        const p = board[m.from];
        return p?.type === 'k' && m.to === toSquare(rank, kingFile);
      }) ?? null
    );
  }

  const cleaned = san.replace(/[+#]/g, '');
  const hasCapture = cleaned.includes('x');
  const hasPromo = cleaned.includes('=');

  let pieceType: PieceType = 'p';
  let targetSq: Square | null = null;
  let disambigFile: number | null = null;
  let disambigRank: number | null = null;
  let promotionType: PieceType | undefined;

  const parts = hasPromo ? cleaned.split('=') : [cleaned];
  const mainPart = parts[0]!;

  if (hasPromo && parts[1]) {
    promotionType = parts[1]!.toLowerCase() as PieceType;
  }

  if (mainPart[0]! >= 'A' && mainPart[0]! <= 'Z') {
    const pieceMap: Record<string, PieceType> = {
      K: 'k',
      Q: 'q',
      R: 'r',
      B: 'b',
      N: 'n',
    };
    pieceType = pieceMap[mainPart[0]!] ?? 'p';
  }

  const withoutPiece = pieceType !== 'p' ? mainPart.slice(1) : mainPart;
  const moveStr = withoutPiece.replace('x', '');

  if (moveStr.length >= 2) {
    const destStr = moveStr.slice(-2);
    targetSq = toSquareFromName(destStr);

    if (moveStr.length > 2) {
      const disambigStr = moveStr.slice(0, -2);
      if (disambigStr.length === 1) {
        const f = 'abcdefgh'.indexOf(disambigStr[0]!);
        if (f >= 0) disambigFile = f;
        else disambigRank = parseInt(disambigStr[0]!) - 1;
      } else if (disambigStr.length >= 2) {
        const sq = toSquareFromName(disambigStr);
        if (sq !== null) {
          disambigFile = getFile(sq);
          disambigRank = getRank(sq);
        }
      }
    }
  }

  if (targetSq === null) return null;

  return (
    moves.find((m) => {
      const p = board[m.from];
      if (!p || p.type !== pieceType || m.to !== targetSq) return false;
      if (promotionType && m.promotion !== promotionType) return false;
      if (hasCapture && !m.captured) return false;
      if (disambigFile !== null && getFile(m.from) !== disambigFile)
        return false;
      if (disambigRank !== null && getRank(m.from) !== disambigRank)
        return false;
      return true;
    }) ?? null
  );
};

File: packages/modules/chess/go/core/notation/notation.go
package notation

import (
	"strconv"
	"strings"

	"github.com/hieudoanm/chess/core/attack"
	"github.com/hieudoanm/chess/core/board"
	"github.com/hieudoanm/chess/core/moves"
	"github.com/hieudoanm/chess/core/types"
	"github.com/hieudoanm/chess/core/utils"
)

func ParseFEN(fen string) types.GameState {
	parts := strings.Split(fen, " ")

	b := board.BoardFromFen(parts[0])

	turn := types.White
	if parts[1] == "b" {
		turn = types.Black
	}

	castlingRights := types.CastlingRights{}
	if parts[2] != "-" {
		for _, ch := range parts[2] {
			switch ch {
			case 'K':
				castlingRights.WK = true
			case 'Q':
				castlingRights.WQ = true
			case 'k':
				castlingRights.BK = true
			case 'q':
				castlingRights.BQ = true
			}
		}
	}

	enPassant := types.NoSquare
	if parts[3] != "-" {
		if sq, ok := utils.ParseSquare(parts[3]); ok {
			enPassant = sq
		}
	}

	halfMoveClock, _ := strconv.Atoi(parts[4])
	fullMoveNumber, _ := strconv.Atoi(parts[5])
	if fullMoveNumber < 1 {
		fullMoveNumber = 1
	}

	inCheck := attack.IsInCheck(b, turn)

	return types.GameState{
		Board:          b,
		Turn:           turn,
		CastlingRights: castlingRights,
		EnPassant:      enPassant,
		HalfMoveClock:  halfMoveClock,
		FullMoveNumber: fullMoveNumber,
		History:        nil,
		Status:         types.StatusPlaying,
		Result:         types.ResultOngoing,
		InCheck:        inCheck,
	}
}

func StringifyFEN(state types.GameState) string {
	fen := board.BoardToFen(state.Board)

	castling := ""
	if state.CastlingRights.WK {
		castling += "K"
	}
	if state.CastlingRights.WQ {
		castling += "Q"
	}
	if state.CastlingRights.BK {
		castling += "k"
	}
	if state.CastlingRights.BQ {
		castling += "q"
	}
	if castling == "" {
		castling = "-"
	}

	ep := "-"
	if state.EnPassant != types.NoSquare {
		ep = utils.SquareName(state.EnPassant)
	}

	return fen + " " + string(state.Turn) + " " + castling + " " + ep + " " +
		strconv.Itoa(state.HalfMoveClock) + " " + strconv.Itoa(state.FullMoveNumber)
}

var pieceLetter = map[types.PieceType]string{
	types.Pawn:   "",
	types.Knight: "N",
	types.Bishop: "B",
	types.Rook:   "R",
	types.Queen:  "Q",
	types.King:   "K",
}

func MoveToUCI(move types.Move) string {
	return utils.SquareName(move.From) + utils.SquareName(move.To) + string(move.Promotion)
}

func ParseUCI(uci string) *types.Move {
	if len(uci) < 4 {
		return nil
	}
	from, ok := utils.ParseSquare(uci[:2])
	if !ok {
		return nil
	}
	to, ok := utils.ParseSquare(uci[2:4])
	if !ok {
		return nil
	}
	var promotion types.PieceType
	if len(uci) >= 5 {
		switch uci[4] {
		case 'q', 'r', 'b', 'n':
			promotion = types.PieceType([]byte{uci[4]})
		}
	}
	return &types.Move{From: from, To: to, Promotion: promotion, Captured: nil}
}

func MoveToSAN(b types.Board, move types.Move, turn types.Color, castlingRights types.CastlingRights, enPassant types.Square) string {
	piece := b[move.From]
	if piece == nil {
		return ""
	}

	fileDiff := abs(utils.FileOf(move.To) - utils.FileOf(move.From))

	if piece.Type == types.King && fileDiff == 2 {
		if utils.FileOf(move.To) > utils.FileOf(move.From) {
			return "O-O"
		}
		return "O-O-O"
	}

	if piece.Type == types.Pawn {
		capture := fileDiff != 0 || move.Captured != nil
		promo := ""
		if move.Promotion != "" {
			promo = "=" + strings.ToUpper(string(move.Promotion))
		}
		file := ""
		if capture {
			file = string(utils.SquareName(move.From)[0])
		}
		capInd := ""
		if capture {
			capInd = "x"
		}
		return file + capInd + utils.SquareName(move.To) + promo
	}

	letter := pieceLetter[piece.Type]

	others := filterOthers(b, move, turn, castlingRights, enPassant)

	disambig := ""
	if len(others) > 0 {
		sameFile := false
		sameRank := false
		for _, m := range others {
			if utils.FileOf(m.From) == utils.FileOf(move.From) {
				sameFile = true
			}
			if utils.RankOf(m.From) == utils.RankOf(move.From) {
				sameRank = true
			}
		}
		if sameFile {
			disambig = utils.SquareName(move.From)
		} else if sameRank {
			disambig = string(utils.SquareName(move.From)[0])
		} else {
			disambig = string(utils.SquareName(move.From)[0])
		}
	}

	capture := ""
	if move.Captured != nil {
		capture = "x"
	}
	return letter + disambig + capture + utils.SquareName(move.To)
}

func filterOthers(b types.Board, move types.Move, turn types.Color, castlingRights types.CastlingRights, enPassant types.Square) []types.Move {
	pieceType := b[move.From].Type
	pseudo := moves.GeneratePseudoLegalMoves(b, turn, castlingRights, enPassant)
	var others []types.Move
	for _, m := range pseudo {
		p := b[m.From]
		if p != nil && p.Type == pieceType && m.To == move.To && m.From != move.From {
			legal := moves.LegalMoves(b, turn, castlingRights, enPassant)
			for _, lm := range legal {
				if lm.From == m.From && lm.To == m.To {
					others = append(others, m)
					break
				}
			}
		}
	}
	return others
}

func ParseSAN(san string, b types.Board, turn types.Color, castlingRights types.CastlingRights, enPassant types.Square) *types.Move {
	movesList := moves.LegalMoves(b, turn, castlingRights, enPassant)

	if san == "O-O" || san == "O-O-O" {
		rank := 0
		if turn == types.Black {
			rank = 7
		}
		kingFile := 6
		if san == "O-O-O" {
			kingFile = 2
		}
		for _, m := range movesList {
			p := b[m.From]
			if p != nil && p.Type == types.King && m.To == utils.Square(rank, kingFile) {
				return &m
			}
		}
		return nil
	}

	cleaned := strings.ReplaceAll(strings.ReplaceAll(san, "+", ""), "#", "")
	hasCapture := strings.Contains(cleaned, "x")
	hasPromo := strings.Contains(cleaned, "=")

	pieceType := types.Pawn
	targetSq := types.NoSquare
	disambigFile := -1
	disambigRank := -1
	var promotionType types.PieceType

	parts := []string{cleaned}
	if hasPromo {
		parts = strings.Split(cleaned, "=")
	}
	mainPart := parts[0]

	if hasPromo && len(parts) > 1 && len(parts[1]) > 0 {
		promotionType = types.PieceType([]byte{strings.ToLower(string(parts[1][0]))[0]})
	}

	if len(mainPart) > 0 && mainPart[0] >= 'A' && mainPart[0] <= 'Z' {
		pieceMap := map[byte]types.PieceType{
			'K': types.King,
			'Q': types.Queen,
			'R': types.Rook,
			'B': types.Bishop,
			'N': types.Knight,
		}
		if pt, ok := pieceMap[mainPart[0]]; ok {
			pieceType = pt
		}
	}

	withoutPiece := mainPart
	if pieceType != types.Pawn {
		withoutPiece = mainPart[1:]
	}
	moveStr := strings.ReplaceAll(withoutPiece, "x", "")

	if len(moveStr) >= 2 {
		destStr := moveStr[len(moveStr)-2:]
		if sq, ok := utils.ParseSquare(destStr); ok {
			targetSq = sq
		}

		if len(moveStr) > 2 {
			disambigStr := moveStr[:len(moveStr)-2]
			if len(disambigStr) == 1 {
				f := strings.IndexByte("abcdefgh", disambigStr[0])
				if f >= 0 {
					disambigFile = f
				} else {
					disambigRank = int(disambigStr[0]-'0') - 1
				}
			} else if len(disambigStr) >= 2 {
				if sq, ok := utils.ParseSquare(disambigStr); ok {
					disambigFile = utils.FileOf(sq)
					disambigRank = utils.RankOf(sq)
				}
			}
		}
	}

	if targetSq == types.NoSquare {
		return nil
	}

	for _, m := range movesList {
		p := b[m.From]
		if p == nil || p.Type != pieceType || m.To != targetSq {
			continue
		}
		if promotionType != "" && m.Promotion != promotionType {
			continue
		}
		if hasCapture && m.Captured == nil {
			continue
		}
		if disambigFile != -1 && utils.FileOf(m.From) != disambigFile {
			continue
		}
		if disambigRank != -1 && utils.RankOf(m.From) != disambigRank {
			continue
		}
		return &m
	}

	return nil
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}
